src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,256 @@
+)
+
+;; Miner veto
+(define-public (veto (proposal-id uint))
+    (let (
+        (cur-vetos (default-to u0 (get vetos (map-get? proposal-vetos { proposal-id: proposal-id }))))
+        (begin-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (expiration-block-height (get expiration-block-height (unwrap!
+            (map-get? vote-confirmed-proposals { proposal-id: proposal-id })
+                (err ERR_VOTE_NOT_CONFIRMED))))
+        (vetoed (default-to false (get vetoed (map-get? exercised-veto { proposal-id: proposal-id,
+                                                                         veto-height: block-height }))))
+        (last-miner (unwrap! (get-block-info? miner-address (- block-height u1))
+            (err ERR_FETCHING_BLOCK_INFO)))
+    )
+
+    ;; a miner can only veto once per block
+    (asserts! (not vetoed) (err ERR_ALREADY_VETOED))
+
+    ;; vetoes can only be cast after the voting period has ended
+    (asserts! (>= burn-block-height begin-block-height) (err ERR_VETO_PERIOD_NOT_STARTED))
+
+    ;; vetoes must be case within the veto period
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_VETO_PERIOD_OVER))
+
+    ;; a miner can only veto if they mined the previous block, proving their hash power
+    (asserts! (is-eq contract-caller last-miner) (err ERR_NOT_LAST_MINER))
+
+    ;; a veto cannot be cast if a proposal has already been miner confirmed
+    (asserts! (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        (err ERR_PROPOSAL_CONFIRMED))
+
+    (map-set proposal-vetos { proposal-id: proposal-id } { vetos: (+ u1 cur-vetos) })
+    (map-set exercised-veto { proposal-id: proposal-id, veto-height: block-height }
+                            { vetoed: true })
+    (ok true))
+)
+
+;; Confirm proposal has reached required vote count
+(define-public (confirm-votes (proposal-id uint))
+    (let (
+        (votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (proposal (unwrap! (map-get? proposals { proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL)))
+        (confirmed-count (var-get confirmed-proposal-count))
+    )
+    (let (
+        (expiration-block-height (get expiration-block-height proposal))
+    )
+
+    ;; confirmation fails if invoked after proposal has expired
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+
+    ;; confirmation fails if the required threshold of votes is not met
+    (asserts! (>= (/ (* votes u100) stx-liquid-supply) REQUIRED_PERCENT_STX_VOTE)
+        (err ERR_INSUFFICIENT_VOTES))
+
+    (map-insert vote-confirmed-proposals { proposal-id: proposal-id }
+        { expiration-block-height: (+ VETO_LENGTH expiration-block-height) })
+
+    (ok true)))
+)
+
+;; Confirm proposal hasn't been vetoed
+(define-public (confirm-miners (proposal-id uint))
```

Comment:
I'm a little confused about these two methods -- `confirm-votes` and `confirm-miners`.  It seems that somebody (can be anybody) must call both `confirm-votes` and, if it success, proceed to then call `confirm-miners`.  `confirm-votes` can be called as soon as there are enough tokens voted, but no later than 2016 blocks after the proposal was created, right?  Once the 2016 blocks have passed, miners then have a 1008-block window in which they can veto a proposal.  If they do not do so, then at any time after the 2016 + 1008 block window, someone (anyone) can call `confirm-miners` to activate the proposal.  Is my understanding here correct?

Reply:
Yeah -- that's the idea. A proposal goes through two phases:

Phase 1: STX holder voting. If the proposal isn't confirmed before expiration block, then the proposal fails. Confirmation requires crossing a threshold of the liquid stx.
Phase 2: Miner vetoing. Once a proposal is stx-voter-confirmed, then miners are able to veto the proposal. Vetos must be applied before the veto period ends. Once the veto period ends, a proposal can be confirmed if fewer than a threshold of vetos were applied.

Splitting this into phases makes it clear to miners when a proposal has gained enough support that it needs to be considered for vetoing. It also limits the amount of time that a miner needs to be issuing veto votes.